jupyter_client/manager.py

```diff
@@ -385,18 +385,24 @@ def interrupt_kernel(self):
                 from .win_interrupt import send_interrupt
                 send_interrupt(self.kernel.win32_interrupt_event)
             else:
-                self.kernel.send_signal(signal.SIGINT)
+                self.signal_kernel(signal.SIGINT)
         else:
             raise RuntimeError("Cannot interrupt kernel. No kernel is running!")
 
     def signal_kernel(self, signum):
-        """Sends a signal to the kernel.
+        """Sends a signal to the process group of the kernel (this
+        usually includes the kernel and any subprocesses spawned by
+        the kernel).
 
         Note that since only SIGTERM is supported on Windows, this function is
         only useful on Unix systems.
         """
         if self.has_kernel:
-            self.kernel.send_signal(signum)
+            try:
+                pgid = os.getpgid(self.kernel.pid)
+                os.killpg(pgid, signum)
+            except Exception:
```

Comment:
Is the "too general" really a problem? I tried the patch on Linux and it works as it should. Since I have no Windows machine to test on, I decided to play it safe and just do `except Exception:`.

> or check directly for hasattr(getpgid/killpg)

Does the fact that those functions exist mean that they will succeed? The problem is that I don't really know the failure modes of these `os.foo()` functions.